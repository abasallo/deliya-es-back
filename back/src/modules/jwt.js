import jwt from 'jsonwebtoken'

const getUserFromToken = token => (token ? jwt.verify(token, process.env.JWT_SECRET).email : '')

const getTokenFromRequest = request => (request.headers.authorization || '').split(' ')[1]

export const getAuthenticatedUserFromRequest = request => getUserFromToken(getTokenFromRequest(request))
