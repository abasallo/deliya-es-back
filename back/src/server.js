import 'core-js/stable'
import 'regenerator-runtime/runtime'

import './modules/dotenv'

import './graphql/sequelize'

import { ApolloServer } from 'apollo-server'

import typeDefs from './graphql/schema'

import resolvers from './graphql/resolvers'

import { getAuthenticatedUserFromRequest } from './modules/jwt'

// TODO - Too much information on Authentication errors is returned, is OK in DEV, but should be 401 in PROD
new ApolloServer({ typeDefs, resolvers, context: ({ req }) => ({ authenticatedUserEmail: getAuthenticatedUserFromRequest(req) }) })
  .listen({ port: process.env.PORT || 4000 })
  .then(({ url }) => console.log(`Server ready at ${url}`))
