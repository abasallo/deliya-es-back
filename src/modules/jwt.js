import jwt from 'jsonwebtoken'

export const getUserFromToken = token => (token ? jwt.verify(token, process.env.JWT_SECRET).email : '')

export const getTokenFromRequest = request => {
  try {
    return request.headers.authorization || '' ? (request.headers.authorization || '').split(' ')[1] : ''
  } catch (e) {
    return ''
  }
}

export const getAuthenticatedUserFromRequest = request => getUserFromToken(getTokenFromRequest(request))
