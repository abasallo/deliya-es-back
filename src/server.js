import './modules/dotenv'

import { initSequelize } from './graphql/sequelize'

import { ApolloServer } from 'apollo-server'

import typeDefs from './graphql/schema'

import resolvers from './graphql/resolvers'

import { getAuthenticatedUserFromRequest } from './modules/jwt'

initSequelize().catch(error => console.error(`Unable to initialize database: ${error}`))

new ApolloServer({
  typeDefs,
  resolvers,
  context: ({ req }) => ({ authenticatedUserEmail: getAuthenticatedUserFromRequest(req) })
})
  .listen({ port: process.env.PORT })
  .then(({ url }) => console.log(`Server ready at ${url}`))
  .catch(error => console.error(`Server startup error: ${error}`))
