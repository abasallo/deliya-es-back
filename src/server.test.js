import './server'

import { createApolloFetch } from 'apollo-fetch'

import jwt from 'jsonwebtoken'

const fetch = createApolloFetch({ uri: 'http://localhost:4000/graphql' })

jest.setTimeout(30000)

test('Login with user and password', () =>
  expect(
    fetch({
      query: 'query($email: String, $password: String) { login(email: $email, password: $password) }',
      variables: { email: 'alvaro@basallo.es', password: 'ojete' }
    })
  ).resolves.toMatchSnapshot({ data: { login: expect.any(String) } }))

test('Login with user and incorrect password', () =>
  expect(
    fetch({
      query: 'query($email: String, $password: String) { login(email: $email, password: $password) }',
      variables: { email: 'alvaro@basallo.es', password: 'incorrect' }
    })
  ).resolves.toMatchSnapshot())

test('Request password recovery', () =>
  expect(
    fetch({
      query: 'query($email: String) { requestPasswordRecoveryUrlOverEmail(email: $email) }',
      variables: { email: 'alvaro@basallo.es' }
    })
  ).resolves.toMatchSnapshot())

test('Request password recovery with nonexistent email', () =>
  expect(
    fetch({
      query: 'query($email: String) { requestPasswordRecoveryUrlOverEmail(email: $email) }',
      variables: { email: 'nonexistent-test@host.tld' }
    })
  ).resolves.toMatchSnapshot())

const createUserQuery = `
mutation($names: String, $surnames: String, $email: String!, $password: String!, $isEmailContactAllowed: Boolean!) {
  addUser(user: { names: $names, surnames: $surnames, email: $email, password: $password, isEmailContactAllowed: $isEmailContactAllowed }) {
    id
    names
    surnames
    email
    password
    isEmailContactAllowed
  }
}`
test('Add user', () =>
  expect(
    fetch({
      query: createUserQuery,
      variables: { names: 'names', surnames: 'surnames', email: 'email', password: 'password', isEmailContactAllowed: true }
    })
  ).resolves.toMatchSnapshot({ data: { addUser: { password: expect.any(String) } } }))

jwt.sign({ email: 'alvaro@basallo.es', date: Date.now() }, process.env.JWT_SECRET, { expiresIn: '10m' })
test('Change password with token', async () =>
  expect(
    fetch({
      query: 'mutation($password: String, $token: String) { changePasswordWithToken(password: $password, token: $token) }',
      variables: {
        password: 'newPassword',
        token: await jwt.sign({ email: 'alvaro@basallo.es', date: Date.now() }, process.env.JWT_SECRET, { expiresIn: '10m' })
      }
    })
  ).resolves.toMatchSnapshot())

test('Change password with wrong, or expired, token', () =>
  expect(
    fetch({
      query: 'mutation($password: String, $token: String) { changePasswordWithToken(password: $password, token: $token) }',
      variables: { password: 'newPassword', token: 'wrong' }
    })
  ).resolves.toMatchSnapshot())
