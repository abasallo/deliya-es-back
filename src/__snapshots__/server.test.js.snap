// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add user 1`] = `
Object {
  "data": Object {
    "addUser": Object {
      "email": "email",
      "id": "2",
      "isEmailContactAllowed": true,
      "names": "names",
      "password": Any<String>,
      "surnames": "surnames",
    },
  },
}
`;

exports[`Change password with token 1`] = `
Object {
  "data": Object {
    "changePasswordWithToken": true,
  },
}
`;

exports[`Change password with wrong, or expired, token 1`] = `
Object {
  "data": Object {
    "changePasswordWithToken": false,
  },
}
`;

exports[`Login with user and incorrect password 1`] = `
Object {
  "data": Object {
    "login": "",
  },
}
`;

exports[`Login with user and password 1`] = `
Object {
  "data": Object {
    "login": Any<String>,
  },
}
`;

exports[`Request password recovery 1`] = `
Object {
  "data": Object {
    "requestPasswordRecoveryUrlOverEmail": true,
  },
}
`;

exports[`Request password recovery with nonexistent email 1`] = `
Object {
  "data": Object {
    "requestPasswordRecoveryUrlOverEmail": false,
  },
}
`;
